File: lib/features/commitments/domain/schedule.dart
import 'package:harvest/core/domain/harvest_day.dart';
import 'package:meta/meta.dart';

/// When a habit is due. Projects are implicitly due daily; to-dos are
/// due on their planned day — schedules only apply to habits.
@immutable
sealed class Schedule {
  const Schedule();

  factory Schedule.fromJson(Map<String, dynamic> json) =>
      switch (json['type'] as String) {
        DailySchedule.kind => const DailySchedule(),
        WeeklySchedule.kind => WeeklySchedule(
          weekdays: (json['weekdays'] as List<dynamic>).cast<int>().toSet(),
        ),
        IntervalSchedule.kind => IntervalSchedule(
          everyDays: json['everyDays'] as int,
          anchorDay: HarvestDay.parse(json['anchorDay'] as String),
        ),
        TimesPerWeekSchedule.kind => TimesPerWeekSchedule(
          times: json['times'] as int,
        ),
        final other => throw ArgumentError('unknown schedule type: $other'),
      };

  Map<String, dynamic> toJson();

  /// Whether the habit is due on [day].
  ///
  /// [doneDaysThisWeek] is the number of distinct days already completed in
  /// [day]'s week — only the flexible times-per-week schedule needs it.
  bool isDueOn(HarvestDay day, {int doneDaysThisWeek = 0});
}

/// Every single day.
final class DailySchedule extends Schedule {
  const DailySchedule();

  static const kind = 'daily';

  @override
  Map<String, dynamic> toJson() => {'type': kind};

  @override
  bool isDueOn(HarvestDay day, {int doneDaysThisWeek = 0}) => true;

  @override
  bool operator ==(Object other) => other is DailySchedule;

  @override
  int get hashCode => kind.hashCode;
}

/// Fixed days of the week (e.g. Mon/Wed/Fri).
final class WeeklySchedule extends Schedule {
  const WeeklySchedule({required this.weekdays});

  static const kind = 'weekly';

  /// [DateTime.monday]..[DateTime.sunday].
  final Set<int> weekdays;

  @override
  Map<String, dynamic> toJson() => {
    'type': kind,
    'weekdays': weekdays.toList()..sort(),
  };

  @override
  bool isDueOn(HarvestDay day, {int doneDaysThisWeek = 0}) =>
      weekdays.contains(day.weekday);

  @override
  bool operator ==(Object other) =>
      other is WeeklySchedule &&
      other.weekdays.length == weekdays.length &&
      other.weekdays.containsAll(weekdays);

  @override
  int get hashCode => Object.hashAllUnordered(weekdays);
}

/// Every X days, counted from an anchor day.
final class IntervalSchedule extends Schedule {
  const IntervalSchedule({required this.everyDays, required this.anchorDay});

  static const kind = 'interval';

  final int everyDays;
  final HarvestDay anchorDay;

  @override
  Map<String, dynamic> toJson() => {
    'type': kind,
    'everyDays': everyDays,
    'anchorDay': anchorDay.key,
  };

  @override
  bool isDueOn(HarvestDay day, {int doneDaysThisWeek = 0}) {
    final distance = anchorDay.daysUntil(day);
    return distance >= 0 && distance % everyDays == 0;
  }

  @override
  bool operator ==(Object other) =>
      other is IntervalSchedule &&
      other.everyDays == everyDays &&
      other.anchorDay == anchorDay;

  @override
  int get hashCode => Object.hash(everyDays, anchorDay);
}

/// X times per week, on whichever days I pick as the week unfolds.
final class TimesPerWeekSchedule extends Schedule {
  const TimesPerWeekSchedule({required this.times});

  static const kind = 'timesPerWeek';

  final int times;

  @override
  Map<String, dynamic> toJson() => {'type': kind, 'times': times};

  @override
  bool isDueOn(HarvestDay day, {int doneDaysThisWeek = 0}) =>
      doneDaysThisWeek < times;

  @override
  bool operator ==(Object other) =>
      other is TimesPerWeekSchedule && other.times == times;

  @override
  int get hashCode => times.hashCode;
}

File: packages/core/src/schedule.ts
import { HarvestDay } from './harvest-day.js';

/**
 * When a habit is due. Projects are implicitly due daily and to-dos on
 * their planned day; schedules only apply to habits.
 *
 * Ported from `apps/mobile/lib/features/commitments/domain/schedule.dart`.
 * The JSON is the one the phone stores in `commitments.schedule_json`.
 */
export type Schedule =
  | { readonly type: 'daily' }
  /** Fixed days of the week, 1 (Monday) through 7 (Sunday). */
  | { readonly type: 'weekly'; readonly weekdays: ReadonlySet<number> }
  /** Every [everyDays] days, counted from [anchorDay]. */
  | { readonly type: 'interval'; readonly everyDays: number; readonly anchorDay: HarvestDay }
  /** [times] a week, on whichever days I pick as the week unfolds. */
  | { readonly type: 'timesPerWeek'; readonly times: number };

export const dailySchedule: Schedule = Object.freeze({ type: 'daily' });

function expectInt(value: unknown, what: string): number {
  if (typeof value !== 'number' || !Number.isInteger(value)) {
    throw new TypeError(`${what} must be an integer`);
  }
  return value;
}

/**
 * `Schedule.fromJson`. Throws on an unknown type or a missing field, as
 * the Dart original does; a schedule that cannot be read is a bug to
 * surface, not a habit to quietly make daily.
 */
export function parseSchedule(json: unknown): Schedule {
  if (typeof json !== 'object' || json === null) throw new TypeError('a schedule is an object');
  const raw = json as Record<string, unknown>;
  switch (raw.type) {
    case 'daily':
      return dailySchedule;
    case 'weekly': {
      if (!Array.isArray(raw.weekdays)) throw new TypeError('weekdays must be a list');
      return {
        type: 'weekly',
        weekdays: new Set(raw.weekdays.map((day) => expectInt(day, 'a weekday'))),
      };
    }
    case 'interval': {
      if (typeof raw.anchorDay !== 'string') throw new TypeError('anchorDay must be a day key');
      return {
        type: 'interval',
        everyDays: expectInt(raw.everyDays, 'everyDays'),
        anchorDay: HarvestDay.parse(raw.anchorDay),
      };
    }
    case 'timesPerWeek':
      return { type: 'timesPerWeek', times: expectInt(raw.times, 'times') };
    default:
      throw new TypeError(`unknown schedule type: ${String(raw.type)}`);
  }
}

/** `Schedule.fromJson(jsonDecode(text))`. */
export function parseScheduleJson(text: string): Schedule {
  return parseSchedule(JSON.parse(text));
}

/** `Schedule.toJson`, with the weekdays sorted as the phone writes them. */
export function scheduleToJson(schedule: Schedule): Record<string, unknown> {
  switch (schedule.type) {
    case 'daily':
      return { type: 'daily' };
    case 'weekly':
      return { type: 'weekly', weekdays: [...schedule.weekdays].sort((a, b) => a - b) };
    case 'interval':
      return { type: 'interval', everyDays: schedule.everyDays, anchorDay: schedule.anchorDay.key };
    case 'timesPerWeek':
      return { type: 'timesPerWeek', times: schedule.times };
  }
}

/**
 * Whether the schedule alone says [day] is due. This is not the rule the
 * field uses; that is `isDueOn` in `due.ts`, which adds the start day,
 * pausing and the other seed types on top.
 *
 * [doneDaysThisWeek] is the number of distinct days already completed in
 * [day]'s week; only the times-per-week schedule reads it.
 */
export function scheduleIsDueOn(schedule: Schedule, day: HarvestDay, doneDaysThisWeek = 0): boolean {
  switch (schedule.type) {
    case 'daily':
      return true;
    case 'weekly':
      return schedule.weekdays.has(day.weekday);
    case 'interval': {
      // Dart throws on a modulo by zero; NaN would quietly say "never".
      if (schedule.everyDays === 0) throw new RangeError('everyDays must not be zero');
      const distance = schedule.anchorDay.daysUntil(day);
      return distance >= 0 && distance % schedule.everyDays === 0;
    }
    case 'timesPerWeek':
      return doneDaysThisWeek < schedule.times;
  }
}
